common/functions/src/scalars/strings/concat.rs

```diff
@@ -0,0 +1,142 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+
+use common_datavalues::prelude::*;
+use common_exception::Result;
+
+use crate::scalars::function_factory::FunctionDescription;
+use crate::scalars::function_factory::FunctionFeatures;
+use crate::scalars::Function;
+
+#[derive(Clone)]
+pub struct ConcatFunction {
+    _display_name: String,
+}
+
+impl ConcatFunction {
+    pub fn try_create(display_name: &str) -> Result<Box<dyn Function>> {
+        Ok(Box::new(ConcatFunction {
+            _display_name: display_name.to_string(),
+        }))
+    }
+
+    pub fn desc() -> FunctionDescription {
+        FunctionDescription::creator(Box::new(Self::try_create))
+            .features(FunctionFeatures::default().deterministic())
+    }
+
+    fn concat_column(lhs: DataColumn, columns: &DataColumnsWithField) -> Result<DataColumn> {
+        if lhs.data_type().is_null() {
+            return Ok(lhs);
+        }
+        if columns.is_empty() {
+            let lhs = lhs.cast_with_type(&DataType::String)?;
+            return Ok(lhs);
+        }
+        let rhs = columns[0].column();
+        if rhs.data_type().is_null() {
+            return Ok(rhs.clone());
+        }
+        let lhs = lhs.cast_with_type(&DataType::String)?;
+        let rhs = rhs.cast_with_type(&DataType::String)?;
+        let array = match (lhs, rhs) {
+            (DataColumn::Array(lhs), DataColumn::Array(rhs)) => {
+                let l_array = lhs.string()?;
+                let r_array = rhs.string()?;
+                DFStringArray::new_from_iter_validity(
+                    l_array
+                        .into_no_null_iter()
+                        .zip(r_array.into_no_null_iter())
+                        .map(|(l, r)| [l, r].concat()),
+                    combine_validities(l_array.inner().validity(), r_array.inner().validity()),
+                )
+            }
+            (DataColumn::Array(l_array), DataColumn::Constant(rhs, len)) => {
+                let l_array = l_array.string()?;
+                let r_array = rhs.to_series_with_size(len)?;
```

Comment:
`rhs` does not need to resize to `r_array`, just get `Option<String>` from `DataValue::String`

Reply:
use `StringArrayBuilder` in this scenario, maybe in this way?
```
(DataColumn::Array(lhs), DataColumn::Constant(rhs, len)) => {
                let l_array = lhs.string()?;
                let r_value = match rhs {
                    DataValue::String(val) => val,
                    _ => None,
                };
                let mut builder = StringArrayBuilder::with_capacity(len);
                let mut iter = l_array.into_iter();
                loop {
                    match iter.next() {
                        Some(val) => {
                            if val.is_none() || r_value.is_none() {
                                builder.append_null();
                            } else {
                                let val = val.unwrap();
                                let val2 = r_value.as_ref().unwrap().as_slice();
                                builder.append_value([val, val2].concat());
                            }
                        }
                        None => break,
                    }
                }
                let array = builder.finish();
                DataColumn::Array(array.into_series())
            }
```

Reply:
If `r_value` is null, just return `DFString::full_null`.

Reply:
resolved.